src/common/transformations/src/transformations/common_optimizations/shared_ops_optimization.cpp

```diff
@@ -45,23 +45,26 @@ class NodeComparingVisitor : public ov::AttributeVisitor {
     void on_adapter(const std::string& name, ValueAccessor<std::shared_ptr<ov::Model>>& adapter) override {
         OPENVINO_THROW_NOT_IMPLEMENTED("Can not compare models");
     };
-    ov::AnyMap get_attributes_map() const {
+
+    ov::AnyMap visit_attributes(const std::shared_ptr<Node>& node) {
+        m_attributes_map.clear();
+        node->visit_attributes(*this);
         return m_attributes_map;
-    };
+    }
 
 private:
     ov::AnyMap m_attributes_map;
 };
 
 bool inputs_from_same_source_or_equal_constants(const std::shared_ptr<Node>& lhs, const std::shared_ptr<Node>& rhs) {
-    if (lhs->get_input_size() != rhs->get_input_size())
-        return false;
     size_t input_size = lhs->get_input_size();
+    if (input_size != rhs->get_input_size())
+        return false;
     for (size_t i = 0; i < input_size; ++i) {
-        if (lhs->input_value(i) == rhs->input_value(i))
+        if (lhs->input_value_raw(i) == rhs->input_value_raw(i))
```

Comment:
What about using 
`get_input_descriptor(i).get_output()` 
it returns reference to output.
From this output the pointer and index can be acquired.

The output descriptor could be expanded to provide raw pointer (similar like input), right now it provides only shared pointer.

I would like  avoid adding the `input_value_raw` to public API as it looks like more internal use in OV and current Node API has a lot of members.

Maybe create local helper as free function to get pair  of node raw pointer and index using node and descriptor API (or comparator without tuple creation), if required it can be moved as part of dev API or public API.